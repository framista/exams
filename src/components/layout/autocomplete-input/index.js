import React, { useState, useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import { FormControl, ListGroup, Form } from 'react-bootstrap';
import { getUniqueExamWithProperty } from '../../../utils/examLogic';

const AutocompleteInput = ({ value, handleChange, data, exams, label }) => {
  const [display, setDisplay] = useState(false);
  const [inputValue, setInputValue] = useState(value);
  const options = getUniqueExamWithProperty(exams, data);
  const wrapperRef = useRef(null);

  useEffect(() => {
    handleChange(inputValue);
    // eslint-disable-next-line
  }, [inputValue]);

  const updateInputValue = (value) => {
    setInputValue(value);
    setDisplay(false);
  };

  const handleKeyPress = (e, value) => {
    if (e.key === 'Enter') {
      updateInputValue(value);
    }
  };

  useEffect(() => {
    window.addEventListener('mousedown', handleClickOutside);
    return () => {
      window.removeEventListener('mousedown', handleClickOutside);
    };
  });

  const handleClickOutside = (event) => {
    const { current: wrap } = wrapperRef;
    if (wrap && !wrap.contains(event.target)) {
      setDisplay(false);
    }
  };

  return (
    <Form.Group ref={wrapperRef}>
      <Form.Label>{label}</Form.Label>
      <FormControl
        type="text"
        onClick={() => setDisplay(!display)}
        onChange={(e) => setInputValue(e.target.value)}
        value={value}
        minLength="5"
        maxLength="50"
        required
      />
      {display && (
        <ListGroup>
          {options
            .filter(
              (option) =>
                option.toLowerCase().indexOf(inputValue.toLowerCase()) !== -1
            )
            .map((value, index) => {
              return (
                <ListGroup.Item
                  key={index}
                  tabIndex={0}
                  onClick={() => updateInputValue(value)}
                  onKeyPress={(e) => handleKeyPress(e, value)}
                >
                  {value}
                </ListGroup.Item>
              );
            })}
        </ListGroup>
      )}
      <Form.Control.Feedback type="invalid">
        Pole nie może być puste
      </Form.Control.Feedback>
    </Form.Group>
  );
};

const mapStateToProps = (state) => ({
  exams: state.exams.allExams,
});
export default connect(mapStateToProps)(AutocompleteInput);
