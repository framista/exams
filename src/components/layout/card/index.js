import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Card, Col } from 'react-bootstrap';
import { AiFillStar, AiOutlineStar } from 'react-icons/ai';
import { formatDate } from '../../../utils/date';
import { getBackgroundCard } from '../../../utils/examLogic';
import ModalForm from '../modal-form';
import { updateExam } from '../../../redux/exams/actions';

const ExamCard = ({ exam, updateExam }) => {
  const [modalFormShow, setModalFormShow] = useState(false);

  const { id, subject, unit, tasks, teacher, date, grade } = exam;

  const saveExam = (exam) => {
    updateExam(exam);
    setModalFormShow(false);
  };
  return (
    <Col sm={6} lg={4} xl={3}>
      <Card
        className="mb-4 text-light shadow-lg"
        onClick={() => setModalFormShow(true)}
        style={{ cursor: 'pointer' }}
        bg={getBackgroundCard(grade, date)}
      >
        <Card.Header>
          <Card.Title className="text-capitalize">{subject}</Card.Title>
          <Card.Subtitle className="text-capitalize">{teacher}</Card.Subtitle>
        </Card.Header>
        <Card.Body>
          <Card.Text className="text-capitalize">{unit}</Card.Text>
          <small>{tasks}</small>
        </Card.Body>
        <Card.Footer className="d-flex justify-content-between">
          <small>{formatDate(date)}</small>
          {grade && (
            <div>
              {new Array(6)
                .fill(0)
                .map((_, i) => i + 1)
                .map((starNumber) =>
                  starNumber <= grade ? (
                    <AiFillStar key={starNumber} />
                  ) : (
                    <AiOutlineStar key={starNumber} />
                  )
                )}
              <small className="ml-1">({grade})</small>
            </div>
          )}
        </Card.Footer>
      </Card>
      <ModalForm
        show={modalFormShow}
        onHide={() => setModalFormShow(false)}
        title="Edytuj sprawdzian"
        saveExam={saveExam}
        exam={exam}
      />
    </Col>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    updateExam: (exam) => dispatch(updateExam(exam)),
  };
};

export default connect(null, mapDispatchToProps)(ExamCard);
