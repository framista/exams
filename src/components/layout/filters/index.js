import React from 'react';
import { connect } from 'react-redux';
import { TiDelete } from 'react-icons/ti';
import { deleteFilter } from '../../../redux/exams/actions';

const Filters = ({ filters, deleteFilter }) => {
  const status = {
    failed: { text: 'Niezaliczone', background: 'danger' },
    future: { text: 'Przysz≈Çe', background: 'info' },
    noResult: { text: 'Nieocenione', background: 'secondary' },
    passed: { text: 'Zaliczone', background: 'success' },
  };
  if (filters.length === 0) {
    return null;
  }
  return (
    <div className="d-flex mb-4 flex-wrap">
      {filters.map((filter) => (
        <div
          className={`bg-${status[filter].background} text-light rounded-pill px-3 py-2 mb-2 mr-2 d-flex align-items-center justify-content-between`}
          key={filter}
        >
          <p className="my-0 mr-1">{status[filter].text}</p>
          <div
            className="d-flex align-items-center btn px-0 mx-0 text-light"
            style={{ fontSize: '1.5rem' }}
            onClick={() => deleteFilter(filter)}
          >
            <TiDelete />
          </div>
        </div>
      ))}
    </div>
  );
};

const mapStateToProps = (state) => ({
  filters: state.exams.filters,
});

const mapDispatchToProps = (dispatch) => ({
  deleteFilter: (filter) => dispatch(deleteFilter(filter)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Filters);
