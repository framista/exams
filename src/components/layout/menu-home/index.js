import React, { useState, Fragment } from 'react';
import { connect } from 'react-redux';
import { ButtonGroup, Button, DropdownButton, Dropdown } from 'react-bootstrap';
import { addFilter, clearFilters } from '../../../redux/exams/actions';
import ModalForm from '../../layout/modal-form';

const MenuHome = ({ filters, addFilter, clearFilters }) => {
  const [modalFormShow, setModalFormShow] = useState(false);
  const handleFilter = (filter) => {
    if (filter === 'all') {
      return clearFilters();
    }
    if (filters.indexOf(filter) === -1) {
      return addFilter(filter);
    }
  };
  return (
    <Fragment>
      <ButtonGroup className="mt-2 mb-4">
        <Button variant="success" onClick={() => setModalFormShow(true)}>
          Nowy sprawdzian
        </Button>
        <DropdownButton
          variant="outline-success"
          as={ButtonGroup}
          title="Filtruj"
          id="bg-nested-dropdown"
          onSelect={(e) => handleFilter(e)}
        >
          <Dropdown.Item eventKey="all">Wszystko</Dropdown.Item>
          <Dropdown.Item eventKey="future">Przysz≈Çe</Dropdown.Item>
          <Dropdown.Item eventKey="noResult">Nieocenione</Dropdown.Item>
          <Dropdown.Item eventKey="failed">Niezaliczone</Dropdown.Item>
          <Dropdown.Item eventKey="passed">Zaliczone</Dropdown.Item>
        </DropdownButton>
        <DropdownButton
          variant="outline-success"
          as={ButtonGroup}
          title="Sortuj"
          id="bg-nested-dropdown"
          onSelect={(e) => console.log(e)}
        >
          <Dropdown.Item eventKey="date">Data</Dropdown.Item>
          <Dropdown.Item eventKey="subject">Przedmiot</Dropdown.Item>
        </DropdownButton>
      </ButtonGroup>
      <ModalForm
        show={modalFormShow}
        onHide={() => setModalFormShow(false)}
        title="Dodaj sprawdzian"
      />
    </Fragment>
  );
};

const mapStateToProps = (state) => ({ filters: state.exams.filters });
const mapDispatchToProps = (dispatch) => {
  return {
    addFilter: (filter) => dispatch(addFilter(filter)),
    clearFilters: () => dispatch(clearFilters()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(MenuHome);
