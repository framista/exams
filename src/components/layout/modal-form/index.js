import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Modal, Button, Form, Alert } from 'react-bootstrap';
import { v4 as uuidv4 } from 'uuid';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import AutocompleteInput from '../../layout/autocomplete-input';

const ModalForm = ({ show, onHide, title, exam, saveExam, deleteExam }) => {
  const [examData, setExamData] = useState(exam);
  const [validated, setValidated] = useState(false);
  const [dateError, setDateError] = useState(false);

  useEffect(() => {
    setExamData({ ...exam });
  }, [show, exam]);

  const handleGrade = (e) => {
    setExamData({ ...examData, grade: e.target.value });
  };

  const validateDateAndGrade = () => {
    return !(
      Number.isInteger(parseInt(examData.grade)) &&
      new Date(examData.date).getTime() > new Date().getTime()
    );
  };

  const handleSubmit = (e) => {
    const form = e.currentTarget;
    if (form.checkValidity() === false) {
      e.preventDefault();
      e.stopPropagation();
    } else {
      const isFormValid = validateDateAndGrade();
      setDateError(!isFormValid);
      if (isFormValid) {
        saveExam(examData);
        setExamData(exam);
        setValidated(true);
      }
    }
    setValidated(true);
  };
  return (
    <Modal
      show={show}
      onHide={onHide}
      backdrop="static"
      size="lg"
      centered
      aria-labelledby="contained-modal-title-vcenter"
    >
      {dateError && (
        <Alert variant="danger" className="my-3 mx-3">
          Nie możesz mieć oceny przed napisaniem <strong>sprawdzianu!</strong>
        </Alert>
      )}
      <Modal.Header closeButton>
        <Modal.Title>{title}</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Form noValidate validated={validated} onSubmit={handleSubmit}>
          <AutocompleteInput
            value={examData.subject}
            handleChange={(value) =>
              setExamData({ ...examData, subject: value })
            }
            data={'subject'}
            label="Przedmiot"
          />
          <AutocompleteInput
            value={examData.teacher}
            handleChange={(value) =>
              setExamData({ ...examData, teacher: value })
            }
            data={'teacher'}
            label="Nauczyciel"
          />
          <Form.Group>
            <Form.Label>Dział</Form.Label>
            <Form.Control
              type="text"
              required
              minLength="5"
              maxLength="50"
              value={examData.unit}
              onChange={(e) =>
                setExamData({ ...examData, unit: e.target.value })
              }
            />
            <Form.Control.Feedback type="invalid">
              Pole powinno mieć 5 - 50 znaków
            </Form.Control.Feedback>
          </Form.Group>
          <Form.Group>
            <Form.Label>Zagadnienia</Form.Label>
            <Form.Control
              as="textarea"
              rows="2"
              required
              minLength="15"
              maxLength="100"
              value={examData.tasks}
              onChange={(e) =>
                setExamData({ ...examData, tasks: e.target.value })
              }
            />
            <Form.Control.Feedback type="invalid">
              Pole powinno mieć 15 - 100 znaków
            </Form.Control.Feedback>
          </Form.Group>
          <Form.Group>
            <Form.Label>Ocena</Form.Label>
            <Form.Control
              as="select"
              onChange={handleGrade}
              value={examData.grade}
            >
              <option value="">Brak</option>
              <option value="1">1</option>
              <option value="2">2</option>
              <option value="3">3</option>
              <option value="4">4</option>
              <option value="5">5</option>
              <option value="6">6</option>
            </Form.Control>
          </Form.Group>
          <Form.Group>
            <Form.Label className="mr-3">Data</Form.Label>
            <DatePicker
              selected={examData.date}
              onChange={(date) => setExamData({ ...examData, date })}
              dateFormat="dd.MM.yyyy"
            />
          </Form.Group>
          <Form.Group className="d-flex justify-content-end">
            {deleteExam && (
              <Button
                variant="danger"
                className="mr-4"
                onClick={() => deleteExam(exam.id)}
              >
                Usuń
              </Button>
            )}
            <Button variant="secondary" className="mr-4" onClick={onHide}>
              Anuluj
            </Button>
            <Button variant="primary" type="submit">
              Zapisz
            </Button>
          </Form.Group>
        </Form>
      </Modal.Body>
    </Modal>
  );
};

ModalForm.propTypes = {
  show: PropTypes.bool.isRequired,
  onHide: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  exam: PropTypes.object.isRequired,
  saveExam: PropTypes.func.isRequired,
};

ModalForm.defaultProps = {
  exam: {
    id: uuidv4(),
    teacher: '',
    subject: '',
    unit: '',
    tasks: '',
    grade: '',
    date: new Date(),
  },
};

export default ModalForm;
