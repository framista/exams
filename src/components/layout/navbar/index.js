import React from 'react';
import { connect } from 'react-redux';
import { Navbar, Nav, Form } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import { toogleTheme } from '../../../redux/theme/actions';

const NavbarComponent = (props) => {
  const { background } = props;
  return (
    <Navbar
      collapseOnSelect
      expand="md"
      bg={background}
      variant="dark"
      className="shadow"
    >
      <Form>
        <Form.Check
          className={`bg-${background}`}
          type="switch"
          id="custom-switch"
          label=""
          onClick={props.toogleTheme}
        />
      </Form>
      <Navbar.Toggle aria-controls="responsive-navbar-nav" />
      <Navbar.Collapse id="responsive-navbar-nav">
        <Nav className="ml-auto">
          <Link className="mr-4 my-2 text-light" to="/">
            Wszystko
          </Link>
          <Link className="mr-4 my-2 text-light" to="/subjects">
            Przedmioty
          </Link>
          <Link className="mr-4 my-2 text-light" to="/about">
            O stronie
          </Link>
        </Nav>
      </Navbar.Collapse>
    </Navbar>
  );
};

const mapStateToProps = (state) => ({
  background: state.theme.background,
});
const mapDispatchToProps = (dispatch) => ({
  toogleTheme: () => dispatch(toogleTheme()),
});

export default connect(mapStateToProps, mapDispatchToProps)(NavbarComponent);
