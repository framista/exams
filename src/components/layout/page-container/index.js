import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

const Div = styled.div`
  min-height: calc(100vh - 56px);
  background: ${(props) => props.background};
  transition: background 0.5s ease-in-out;
  padding: 2rem;
`;

const PageContainer = ({ background, children }) => {
  const pageColor = background === 'primary' ? '#F8F9FA' : '#282b2f';
  return <Div background={pageColor}>{children}</Div>;
};

const mapStateToProps = (state) => ({
  background: state.theme.background,
});

export default connect(mapStateToProps)(PageContainer);
