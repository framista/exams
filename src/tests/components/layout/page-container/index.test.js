import React from 'react';
import { shallow } from 'enzyme';
import { PageContainer } from '../../../../components/layout/page-container';

describe('test PageContainer', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<PageContainer />);
    expect(wrapper).toMatchSnapshot();
  });
});
