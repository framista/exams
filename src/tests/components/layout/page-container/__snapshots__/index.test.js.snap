// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PageContainer should render correctly 1`] = `
<styled.div
  background="#282b2f"
/>
`;
